File: backend/src/main/java/com/adp/jobhunter/service/OffreService.java
package com.adp.jobhunter.service;

import com.adp.jobhunter.entity.Offre;
import com.adp.jobhunter.repository.OffreRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

/**
 * SERVICE — OffreService.java
 *
 * C'est la couche logique métier — elle fait le lien entre
 * le Controller (HTTP) et le Repository (base de données).
 *
 * Elle ne sait pas d'où vient la requête HTTP.
 * Elle ne sait pas comment les données sont stockées.
 * Elle sait juste QUOI faire avec les données.
 *
 * Equivalent de la couche services/ dans FastAPI.
 */
@Service
public class OffreService {

    @Autowired  // Spring injecte automatiquement le Repository
    private OffreRepository offreRepository;
    // Equivalent du "db: Session = Depends(get_db)" en FastAPI

    // Récupérer toutes les offres
    public List<Offre> getAll() {
        return offreRepository.findAll();
        // → SELECT * FROM offres
    }

    // Récupérer une offre par son id
    public Optional<Offre> getById(Long id) {
        return offreRepository.findById(id);
        // → SELECT * FROM offres WHERE id = ?
    }

    // Créer ou mettre à jour une offre
    public Offre save(Offre offre) {
        return offreRepository.save(offre);
        // → INSERT ou UPDATE automatiquement
    }

    // Supprimer une offre
    public void delete(Long id) {
        offreRepository.deleteById(id);
        // → DELETE FROM offres WHERE id = ?
    }

    // Rechercher par entreprise
    public List<Offre> getByEntreprise(String entreprise) {
        return offreRepository.findByEntreprise(entreprise);
        // → SELECT * FROM offres WHERE entreprise = ?
    }

    // Rechercher par mot clé dans le titre
    public List<Offre> search(String keyword) {
        return offreRepository.findByTitreContainingIgnoreCase(keyword);
        // → SELECT * FROM offres WHERE LOWER(titre) LIKE LOWER('%keyword%')
    }

    // Rechercher par niveau
    public List<Offre> getByNiveau(String niveau) {
        return offreRepository.findByNiveau(niveau);
        // → SELECT * FROM offres WHERE niveau = ?
    }

    // Rechercher par catégorie
    public List<Offre> getByCategorie(String categorie) {
        return offreRepository.findByCategorie(categorie);
        // → SELECT * FROM offres WHERE categorie = ?
    }
}

File: frontend/src/app/services/offre.service.ts
/**
 * SERVICE — offre.service.ts
 *
 * C'est la couche qui appelle l'API Spring Boot.
 * Equivalent de la couche services/ en FastAPI.
 *
 * HttpClient = requests en Python
 * Observable = async/await en Python
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Offre } from '../models/offre.model';

@Injectable({
  providedIn: 'root'  // disponible partout dans l'app
})
export class OffreService {

  // URL de base de ton API Spring Boot
  private apiUrl = 'http://localhost:8080/api/offres';

  constructor(private http: HttpClient) {}
  // HttpClient injecté automatiquement — comme Depends() en FastAPI

  // GET /api/offres → toutes les offres
  getAll(): Observable<Offre[]> {
    return this.http.get<Offre[]>(this.apiUrl);
  }

  // GET /api/offres/{id} → une offre
  getById(id: number): Observable<Offre> {
    return this.http.get<Offre>(`${this.apiUrl}/${id}`);
  }

  // POST /api/offres → créer une offre
  create(offre: Offre): Observable<Offre> {
    return this.http.post<Offre>(this.apiUrl, offre);
  }

  // GET /api/offres/search?keyword= → rechercher
  search(keyword: string): Observable<Offre[]> {
    return this.http.get<Offre[]>(`${this.apiUrl}/search?keyword=${keyword}`);
  }
}